### DISCUSSION TITLE
Nutzung der weewx-DWD Daten

### DISCUSSION THREAD

**Discussion by hoetzgit** [roe-dl/weewx-DWD#35] [General]
Für die aktuellen Bedingungen (Icon, Text ...). In der Belchertown skin wollte ich mir für den Anfang ein JSON Template mit den Daten `icon` und `icontitle` und ein paar Werten generieren. Dazu wollte ich bspw. `$current.om_Icon.raw` oder `$current.om_Snow.raw` nutzen. Offensichtlich stehen aber die (zumindest manche) `raw` Daten per `$current`/`$latest` ... nicht zur Verfügung. Bei mir zumindest trifft das auch auf die POI/CDC Daten zu. Hab ich da was verhunzt? Dann habe ich überlegt, dass mir das mit JSON nicht gefällt und wollte die Daten lieber in den MQTT Broker schicken. Kann man einen (dann eben eigenen) Fork von `weewx-mqtt` so modifizieren, dass man auf den Inhalt der `current` Objekte zugreifen kann und die Textfelder auch mit in den Broker sendet? Dann könnte ich auch die Textfelder der Skin per MQTT und jQuery updaten. Oder eben über alternative Methoden, aber im Prinzip hätte ich gerne auch Textfelder im Broker

**Comment by roe-dl** [comment]
Alle die Daten, die über `dwd.py` in den Datensatz eingefügt werden, sollten auch mittels `$current` zur Verfügung stehen, gleich ob mit `.raw` oder ohne. Wenn nicht, sollten irgendwo auch Fehlermeldungen zu sehen sein.

Bei `$latest` ist es etwas anderes. Dort sind nur diejenigen Daten verfügbar, die auch gespeichert werden. 

MQTT erlaubt auch Text. Dein Gedanke sollte machbar sein.

**Reply by hoetzgit** [reply]
> Alle die Daten, die über `dwd.py` in den Datensatz eingefügt werden, sollten auch mittels `$current` zur Verfügung stehen, gleich ob mit `.raw` oder ohne. Wenn nicht, sollten irgendwo auch Fehlermeldungen zu sehen sein.

die Fehlermeldung kommt (ich habe alle Felder zur Sicherheit auch in der `extensions.py` den Gruppen zugeordnet.)

```
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: **** Ignoring template /home/weewx/skins/Weiherhammer/json/weewx_dwd.json.tmpl
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: **** Reason: 'UnknownType' object is not subscriptable
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****  Traceback (most recent call last):
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****    File "/home/weewx/bin/weewx/cheetahgenerator.py", line 348, in generate
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****      unicode_string = compiled_template.respond()
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****    File "_home_weewx_skins_Weiherhammer_json_weewx_dwd_json_tmpl.py", line 105, in respond
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****    File "_home_weewx_skins_Weiherhammer_json_weewx_dwd_json_tmpl.py", line 72, in __errorCatcher1
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****    File "<string>", line 1, in <module>
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****    File "/home/weewx/bin/weewx/units.py", line 1049, in raw
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****      return self.value_t[0]
Jan 31 11:56:11 weewx wee_reports[2240290] ERROR weewx.cheetahgenerator: ****  TypeError: 'UnknownType' object is not subscriptable

```
Hier das template, ist nur ein Muster. bspw. hier `$current.om_DateTime.raw`

```
#errorCatcher Echo
#encoding UTF-8
#import datetime
{
    "timestamp": #echo int( time.time() )#,
    "openmeteo_current": {
        "timestamp": $current.om_DateTime.raw,
        "current": [
            {
                "success": true,
                "error": null,
                "source": "Open-Meteo (DWD Daten)",
                "response": {
                    "dataSource": "dwd.py",
                    "loc": {
                        "latitude": 49.632270,
                        "longitude": 12.056186,
                        "altitudeM": 394
                    },
                    "place": {
                        "name": "Weiherhammer",
                        "city": "Weiherhammer",
                        "state": "by",
                        "country": "de"
                    },
                    "profile": {
                        "tz": "Europe/Berlin",
                        "tzname": "GMT",
                        "tzoffset": 0
                    },
                    ##"obTimestamp": $current.om_DateTime.raw,
                    "ob": {
                        ##"timestamp": $current.om_DateTime.raw,
                        ##"apptempC": $current.om_AppTemp.raw,
                        ##"barometerMB": $current.om_Barometer.raw,
                        ##"dewpointC": $current.cdc_Dewpoint.raw,
                        ##"et": $current.om_Et.raw,
                        ##"FreezinglevelHeight ": $current.om_FreezinglevelHeight .raw,
                        ##"humidity": $current.om_OutHumidity.raw,
                        ##"icon": "$current.om_Icon.raw",
                        ##"radiationWM2": $current.om_Radiation.raw,
                        ##"rain": $current.om_Rain.raw,
                        ##"shower": $current.om_Shower.raw,
                        ##"sky": $current.om_Cloudcover.raw,
                        ##"snow": $current.om_Snow.raw,
                        ##"snowdepth": $current.om_SnowDepth.raw,
                        ##"tempC": $current.om_OutTemp.raw,
                        ##"weather": "$current.om_Icontitle.raw",
                        ##"weathercode": $current.om_Weathercode.raw,
                        ##"weatherPrimary": "$current.om_Icontitle.raw",
                        ##"windDirDEG": $current.om_WindDir.raw,
                        ##"windGustDirDEG": $current.om_WindGustDir.raw,
                        ##"windGustKPH": $current.om_WindGust.raw,
                        ##"windSpeedKPH": $current.in_WindSpeed.raw,
                        ####if $current.om_snowfallHeight.has_data
                        ##"snowfallHeight ": $current.om_snowfallHeight.raw,
                        ####else
                        "snowfallHeight ": null,
                        ####end if
                        "isDay": false
                    }
                }
            }
        ]
    },
##TODO
    "dwd_cdc": {
##        "timestamp": $latest.cdc_DateTime.raw,
        "current": [
            {
                "success": true,
                "error": null,
                "source": "DWD CDC Daten",
                "response": {
                    "dataSource": "dwd.py",
                    "loc": {
                        "latitude": 49.632270,
                        "longitude": 12.056186,
                        "altitudeM": 394
                    },
                    "place": {
                        "name": "Weiherhammer",
                        "city": "Weiherhammer",
                        "state": "by",
                        "country": "de"
                    },
                    "profile": {
                        "tz": "Europe/Berlin",
                        "tzname": "GMT",
                        "tzoffset": 0
                    },
##                    "obTimestamp": $latest.cdc_DateTime.raw,
                    "ob": {
                    }
                }
            }
        ]
    },
##TODO
    "dwd_poi": {
    }
}
```

**Reply by roe-dl** [reply]
Die Fehlermeldung sagt normalerweise, daß irgendein Wert `None` ist und als Array oder Dictionary benutzt wird. In dem Fall würde ich vorschlagen, klein anzufangen, also mit den geschweiften Klammern am Anfang und Ende, und dann Wert für Wert hinzuzufügen, bis der Fehler kommt. Dann wissen wir wenigstens, welche Variable ihn erzeugt.

**Reply by roe-dl** [reply]
Man kann `dwd.py` auch allein ohne WeeWX aufrufen. Dazu braucht man eine Datei namens `DWD.conf` im aktuellen Verzeichnis. Diese Datei muß die nötigen Angaben aus `weewx.conf` enthalten (kann auch eine Kopie von `weewx.conf` sein). Dann ruft man die Datei mit 
```python
python3 /usr/share/weewx/user/dwd.py
```
auf und erhält in der Terminalausgabe die Werte, die normalerweise hinzugefügt würden. Das kann die Fehlersuche auch erleichtern. Möglicherweise ist ja ein Kandidat dabei.

**Reply by roe-dl** [reply]
Es gibt auch 
```
$station.altitude
$station.latitude
$station.longitude
$station.location
$station.stn_info.latitude_f
$station.stn_info.longitude_f
```

**Reply by roe-dl** [reply]
Ich verstehe auch nicht, warum die Werte in diesem Zusammenhang alle mit `.raw` am Ende angegeben sind. Das macht man eigentlich nur bei Variablenzuweisungen und Stringvariablen. `dateTime` ist aber keine Stringvariable sondern eine Zahlenvariable.

**Reply by hoetzgit** [reply]
> Ich verstehe auch nicht, warum die Werte in diesem Zusammenhang alle mit `.raw` am Ende angegeben sind. Das macht man eigentlich nur bei Variablenzuweisungen und Stringvariablen. `dateTime` ist aber keine Stringvariable sondern eine Zahlenvariable.

Schuld ist - wie immer - das oft falsche Icon in der Belchertown skin. Nachdem Du `dwd-mosmix` erstellt hast, war das schon eine Verbesserung. Dann habe ich auf der Suche nach einem bestmöglichen Ergebnis angefangen, `dwd-mosmix` zu missbrauchen und weitere Wetterdienste abzufragen, Flughafen API, OWM, Brightsky etc um aktuelle Bedingungen zu bekommen. Da meine Belchertown skin eh schon ein Fork ist, habe ich in der `forecast.json` für jeden Dienst einen eigenen Abschnitt eingeführt. Auf meiner Seite kann ich dadurch unter dem Icon die Dienste zum Test einfach durchschalten und vergleichen. Der Einfachheit halber wollte ich nun noch die CDC/POI/Open-Meteo Ergebnisse in gleicher Form in ein JSON Template packen (Beispiel oben). Javascript der Skin setzt die zwei Dateien zusammen und ich kann dann auch die drei zusätzlich Dienste durchschalten. Ich brauche also strenggenommen nur raw Werte für `cloudcover`, wenn vorhanden `visibility`, hauptsächlich aber Icon und Wetterbeschreibung in dem JSON Template. 

In Zukunft möchte ich ja alle dynamischen Daten im Broker haben. Egal ob Werte, HTML Segmente, JSON was auch immer. Per MQTT und jQuery mache ich dann die entsprechenden Updates auf der Seite, auch Vorhersageelemente, etc. Nur bei einem MQTT Problem möchte ich alle diese Werte aus einer im Archiv Intervall erstellten JSON lesen und die Seite updaten. Deshalb habe ich mich schon mal auf `.raw` eingeschossen. `$current.dateTime` ohne `.raw` erzeugt übrigens "31.01.2023 15:50:00". Und auch bei Zahlenvariablen macht `.raw` Sinn, ich möchte ja kein Komma als Dezimaltrennzeichen, sondern einen Punkt und auch die komplette Nachkommastellenanzahl.

Während ich hier schreibe kam mir die Idee, dass das Problem mit dem Fehler von oben daran liegen könnte, das ich zum Test der Templates immer `wee_reports` gestartet habe. Und tatsächlich, werden die Reports im Archiv Intervall von WeeWX gestartet, sind alle Werte da und die Templates werden korrekt erstellt. Den Text drüber lasse ich jetzt aber ;-)

**Comment by hoetzgit** [comment]
Kann es sein, dass für CDC keine Werte für `icon` und `icontitle` generiert werden? Dann wundert es mich nicht, dass ich immer Fehler bekomme, wenn ich z.B. `$current.cdc_Icon.raw` benutze.

**Reply by hoetzgit** [reply]
> Kann es sein, dass für CDC keine Werte für `icon` und `icontitle` generiert werden? Dann wundert es mich nicht, dass ich immer Fehler bekomme, wenn ich z.B. `$current.cdc_Icon.raw` benutze.

Frage gerade dank Doku selbst beantwortet. Sorry
